velox/exec/prefixsort/PrefixSortEncoder.h

```diff
@@ -228,4 +237,126 @@ FOLLY_ALWAYS_INLINE void PrefixSortEncoder::encodeNoNulls(
   encodeNoNulls(value.getNanos(), dest + 8);
 }
 
+class PrefixSortLongDecimalToIntEncoder : public PrefixSortEncoder {
+ public:
+  PrefixSortLongDecimalToIntEncoder(
+      bool ascending,
+      bool nullsFirst,
+      int precision,
+      int scale)
+      : PrefixSortEncoder(ascending, nullsFirst),
+        precision_(precision),
+        scale_(scale),
+        newScale_(newPrecision_ - (precision_ - scale_)) {
+    VELOX_DCHECK(precision - scale <= ShortDecimalType::kMaxPrecision);
+  };
+
+  FOLLY_ALWAYS_INLINE void encode(std::optional<int128_t> value, char* dest)
+      const override {
+    if (value.has_value()) {
+      dest[0] = nullsFirst_ ? 1 : 0;
+      int64_t newValue;
+      auto status = DecimalUtil::rescaleWithRoundUp<int128_t, int64_t>(
+          value.value(),
+          precision_,
+          scale_,
+          newPrecision_,
+          newScale_,
+          newValue);
+      if (FOLLY_UNLIKELY(!status.ok())) {
+        if (value < 0) {
+          newValue = std::numeric_limits<int64_t>::min();
+        } else {
+          newValue = std::numeric_limits<int64_t>::max();
+        }
+      }
+      encodeNoNulls(newValue, dest + 1);
+    } else {
+      dest[0] = nullsFirst_ ? 0 : 1;
+      simd::memset(dest + 1, 0, encodedSize() - 1);
+    }
+  }
+
+  static int32_t encodedSize() {
+    return 9;
+  }
+
+ private:
+  const int precision_;
+  const int scale_;
+  const int newPrecision_ = ShortDecimalType::kMaxPrecision;
+  const int newScale_;
+};
+
+class PrefixSortHugeIntEncoder : public PrefixSortEncoder {
+ public:
+  PrefixSortHugeIntEncoder(
+      bool ascending,
+      bool nullsFirst,
+      int precision,
+      int scale)
+      : PrefixSortEncoder(ascending, nullsFirst),
+        precision_(precision),
+        scale_(scale){};
+
+  FOLLY_ALWAYS_INLINE void encode(std::optional<int128_t> value, char* dest)
+      const override {
+    if (value.has_value()) {
+      dest[0] = nullsFirst_ ? 1 : 0;
+      auto val = value.value();
+      encodeNoNulls<int64_t>(HugeInt::upper(val), dest + 1);
```

Comment:
I think we should not make semantic changes here in sort？Can you prove that this is OK under any circumstances? Or is there any relevant document reference?

Reply:
https://github.com/duckdb/duckdb/blob/main/src/include/duckdb/common/radix.hpp#L225 duckdb implementation